Run PHPUnit jobs across multiple PHP versions

anton-vlasenko opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR:
- Removes PHP related checks from the `Unit Tests` Github CI job;
- Adds 2 new Github CI Jobs: `PHPUnit Tests` and `PHP Coding Standards` (the same way it's implemented in Core);
- **brings ability to run PHPUnit tests on all PHP versions (5.6 through 8.2) using the CI jobs;**
- shows code style errors on the "Files changed" tab (same as Core);
- Fixes an issue with `wp-env` as it used non-optimal PHPUnit versions on PHP 7.3 and 7.4 (props to @jrfnl);

## Why?
Gutenberg should support the same PHP versions as Core.

## How?
This PR adds 2 new Github CI Jobs: `PHPUnit Tests` and `PHP Coding Standards`.
These PHP CI jobs were backported from Core to ensure that we test Gutenberg PRs in the same fashion.

## Testing Instructions
1. Make sure that the CI jobs run without errors.
2. Observe there are `PHP_VERSION (multisite) on ubuntu-latest` jobs for each supported PHP version (5.6 through 8.2).

Props @jrfnl , @dmsnell , @anton-vlasenko

anton-vlasenko committed:
Add a small change to create a PR.

anton-vlasenko committed:
Add multiple php versions.

anton-vlasenko committed:
Add multiple php versions.

anton-vlasenko committed:
Define phpVersion as environment variable.

anton-vlasenko committed:
Add phpunit tags for PHP 8.1 and 8.2.

anton-vlasenko committed:
Remove hyphen.

anton-vlasenko committed:
Nove PHPUnit tests to a separate file.

anton-vlasenko committed:
Disable `fail-fast`.

anton-vlasenko committed:
Update matrix.os.

anton-vlasenko committed:
Install composer dependencies.

anton-vlasenko committed:
Revert "Install composer dependencies."

This reverts commit e433d838d3c6b1a2e546ac407647377e3273050d.

anton-vlasenko committed:
Revert "Revert "Install composer dependencies.""

This reverts commit 5e13c24e36aec30b69f3a9ab71d06833bdd4ad4b.

aaroncampbell committed:
Fixing "no phpunit found" error.

anton-vlasenko committed:
Revert "Fixing "no phpunit found" error."

This reverts commit ee422158de53dd7525009338013fdfcbd0c6c9c6.

anton-vlasenko committed:
Add phpunit tags for PHP 8.1 and 8.2.

anton-vlasenko committed:
Debug the phpunitImage.

anton-vlasenko committed:
Use correct composer version.

anton-vlasenko committed:
Fix container name.

anton-vlasenko committed:
Don't modify package.json.
Run tests in phpunit-test.yml instead.

anton-vlasenko committed:
Don't install composer dependencies twice.

anton-vlasenko committed:
Need to prefix commands with npm run.

anton-vlasenko committed:
Add double dashes to pass flags to wp-env.

anton-vlasenko committed:
Fix name of the environment.

anton-vlasenko committed:
Add additional debug information.

anton-vlasenko committed:
Composer dependencies should be installed after wp-env is intialised.

anton-vlasenko committed:
Fix wrong command.

anton-vlasenko committed:
Since the tests are currently run within the Docker containers where the PHP version varies, the same PHP version needs to be configured for the action runner.

anton-vlasenko committed:
Fix path to the multisite.xml.

anton-vlasenko committed:
Fix the test class to be compatible with PHP < 7.3.

anton-vlasenko committed:
Fix bug in WP_REST_Block_Pattern_Categories_Controller_Test::wpTearDownAfterClass.

anton-vlasenko committed:
Remove incompatible PHP packages.

anton-vlasenko committed:
Remove incompatible PHP packages.

anton-vlasenko committed:
Update description.

anton-vlasenko committed:
Remove incompatible PHP packages.

anton-vlasenko committed:
Remove incompatible PHP packages.

anton-vlasenko committed:
Fix wrong matrix value.

anton-vlasenko committed:
Remove incompatible PHP packages.

anton-vlasenko committed:
Don't remove polyfills.

anton-vlasenko committed:
Don't remove spathie package on PHP 7.2.

anton-vlasenko committed:
// strpos() doesn't support negative offset values on PHP prior to version 7.1.

anton-vlasenko committed:
Add cliPhpVersion config parameter.

anton-vlasenko committed:
Don't use custom commands to run the tests.

anton-vlasenko committed:
Revert "Add cliPhpVersion config parameter."

This reverts commit 1a73cd90908a85711c7cc9e7c567a7dc3528b0fa.

anton-vlasenko committed:
Don't use the phpunit docker-compose service.

anton-vlasenko committed:
Fix dynamic properties. PHP 8.2 doesn't support dynamic properties.

anton-vlasenko committed:
Fix dynamic properties. PHP 8.2 doesn't support dynamic properties.

anton-vlasenko committed:
Fix dynamic properties. PHP 8.2 doesn't support dynamic properties.

anton-vlasenko committed:
Allow /.cache/.gitkeep files.

anton-vlasenko committed:
Add the .gitkeep file.
The .cache directory is needed for CI/build tools.

anton-vlasenko committed:
Add PHP Coding Stadnards.

anton-vlasenko committed:
Fix the if condition.

anton-vlasenko committed:
Fix path.

anton-vlasenko commented on .gitignore:
```diff
@@ -17,6 +17,9 @@ yarn.lock
 /composer.lock
 
 .cache
+!/.cache/
```
This ignores all `.gitkeep` files but `/.caches/.gitkeep`.
The `/.caches/` folder is needed for the `PHP Coding Standards` CI job.

anton-vlasenko replied:
This ignores all `.gitkeep` files but `/.caches/.gitkeep`.
The `/.caches/` folder is needed for the `PHP Coding Standards` CI job.

anton-vlasenko commented on .gitignore:
```diff
@@ -17,6 +17,9 @@ yarn.lock
 .cache
+!/.cache/
+/.cache/**
+!/.cache/.gitkeep
```
This ignores all `.cache/.gitkeep` files but `/.cache/.gitkeep`.
The `/.cache/` folder is needed for the `PHP Coding Standards` CI job.

dmsnell replied:
This deserves an explanatory comment inline as it's going to be really hard to find this PR comment when someone is trying to understand why the `.cache` directory is so zig-zagged here. Can we add the explanation, and include why a simpler formulation of the rules cannot provide what we need?

Is it not possible to have `PHP Coding Standards` use another directory? What is the reason for its use of `.cache` here anyway?

dmsnell replied:
I also now see that we're telling `git` to keep the `.cache` directory, which seems surprising to me. why do we need to do this?

if this weren't `.gitkeep`'d then this surprising change to `.gitignore` presumably would also not be necessary

anton-vlasenko replied:
The reason for zig-zagging [is the following](https://git-scm.com/docs/gitignore):
```
It is not possible to re-include a file if a parent directory of that file is excluded. 
```

> This deserves an explanatory comment inline as it's going to be really hard to find this PR comment when someone is trying to understand why the .cache directory is so zig-zagged here.

I agree. Fixed in 89a4d738e13c7d6ff51d54cfa2477eb0c713c67b

Currently, Gutenberg ignores all `.cache` folders (regardless of their location).
For compatibility reasons, these folders should still be ignored, while allowing the /.cache folder (in the root).
I haven't been able to come up with a simpler pattern that would allow that.

> if this weren't .gitkeep'd then this surprising change to .gitignore presumably would also not be necessary

Git doesn't allow to commit empty folders. If it did, I would commit an empty folder.

> Is it not possible to have PHP Coding Standards use another directory? What is the reason for its use of .cache here anyway?

Yes, it's possible, but I don't understand what's wrong with `/.cache`.

A similar Core job [uses a folder with](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/coding-standards.yml#L88) the same name -  `.cache`.

dmsnell replied:
sure but why do we need or want to commit an empty directory?

jrfnl replied:
Because PHPCS expects the directory to which to write the cache file to exist. It will not create it.

So if the ruleset dictates that the cache file should be written to the `.cache` directory, that directory _must_ exist as otherwise PHPCS will error out.
You could, of course, choose to just let the cache file be written to the system temp directory or to the project root...

Note: the PHPCS cache file itself should definitely be `gitignored`.

dmsnell replied:
my suspicion was that it was something like this, but if that's the case, that doesn't explain why it needs to be in `git`, it merely notes that the directory needs to be present when running the linter.

if the entire purpose of coercing `git` to include an empty directory in the repo is so that a linter can run in CI then it would be simpler to create the directory when running the CI script with `mkdir -p .cache` or the equivalent.

jrfnl replied:
Well, that depends. If the ruleset contains the `cache` directive, local CS runs will also expect the directory.

If the caching is only used in CI, I agree, a `mkdir` would suffice, but then again, why would the caching be limited to CI ?

Also note - in most projects which use a `.cache` directory for the PHPCS cache placement, it is generally because they want to cache those files in CI and Travis (yes, that's the completely outdated reason), didn't allow for caching individual files, only allowed for caching directories, which is why people resorted to having a `.cache` directory for this.

GH Actions, however, _does_ allow for caching individual files, so I honestly don't know why that `.cache` directory is still used, especially for projects setting this up from scratch. (I can understand people not having revisited their previous reason for setting it up when converting from Travis to GHA, but that's not the case here)

dmsnell commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+            # http://man7.org/linux/man-pages/man1/date.1.html
+            - name: "Get last Monday's date"
+              id: get-date
+              run: echo "date=$(/bin/date -u --date='last Mon' "+%F")" >> $GITHUB_OUTPUT
```
Can we get an explanation of why we want to clear the cache once a week?
Generally time-based caches are useful in rare and somewhat emergency situations.

What is the reason we can't use a content-based cache?
Also do we know we need the cache? What do we expect it to provide for us?

anton-vlasenko replied:
This cache is both content and time-based.
```
hashFiles('**/composer.json', 'phpcs.xml.dist')
```
ensures that the cache key is content-based.

> Also do we know we need the cache? What do we expect it to provide for us?

The main benefit of using the cache is to speed up the CI jobs.

dmsnell replied:
What have you measured as the speedup in the CI jobs system?

anton-vlasenko replied:
In my tests,
```
phpcs -n --report-full 
```
takes about 12 seconds when the cache is disabled and 0 seconds (that's what GitHub shows) when the cache is enabled.
As the quantity of `.php` files in the project is growing, this difference will only increase over time.

dmsnell replied:
hm. what if we introduced this first without the cache? in the grander scheme of all the test suites 12s is a drop in the bucket but the cache is going to likely create a non-trivial amount of headache at the point it causes problems in a PR.

I don't think these tests are going to be the slowest workflow suite, which means they aren't in the hot path, and if we merge first without the cache then if we think one is truly warranted then we can make a PR and gather a statistical measurement to determine if, when put into the system with 1700 open branches and lots of CI activity, if it still makes 12s of an improvement and if we think that's worth the additional complexity

jrfnl replied:
> Can we get an explanation of why we want to clear the cache once a week?
> What is the reason we can't use a content-based cache?

Without having looked at the implementation in this particular workflow...
* Caches in GitHub Actions work differently than people often expect from other CI systems.
* A cache in GHA is created based on a key and NEVER updated.
* The only way to vacate the cache is to create a new cache with a new key.

So, for caches where the key doesn't change often, the cache tends to go more and more out of date over time. Adding a time based factor to the key helps to vacate the cache regularly.

Also see: https://github.com/ramsey/composer-install/issues/234

> hm. what if we introduced this first without the cache?

In case it would be decided to go that way, I would **_strongly_** recommend adding the `--no-cache` CLI argument to the PHPCS command in GHA.
As the `cache` directive is in the ruleset, PHPCS would otherwise needlessly create the cache file from scratch each time, which has a significant impact on the runtime of PHPCS due to the file-write operations.
As CLI args overrule the ruleset, adding `no-cache` should already speed up the CS run.

dmsnell commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,128 @@
+            # It must be removed before running the tests.
+            - name: Remove incompatible Composer packages
+              if: ${{ matrix.php < '7.2' }}
+              run: composer remove spatie/phpunit-watcher --dev --no-update
```
if this breaks in supported PHP versions, why not remove the dependency altogether? I'm worried about the disconnect between requiring this package and having this line in the workflow YML. best case scenario is that when this dependency is no longer required, we'll still be running this step, which is benign, but also is going to confuse someone why it's there.

is there a way to have composer tell us which packages are not compatible? or should we file a bug report with `spatie/phpunit-watch` to have that project list a minimum version if one is missing?

anton-vlasenko replied:
> if this breaks in supported PHP versions, why not remove the dependency altogether?

This is because `spatie/phpunit-watcher` is required by the `npm run test:php:watch` command.
Thus it cannot be simply removed.
I was going to create a new GB issue to discuss the possibility of replacing it with a similar package that supports PHP < 7.2 (if that substitute package exists).

> is there a way to have composer tell us which packages are not compatible?

Yes, composer already does that. It knows that this package is not compatible with PHP < 7.2 and refuses to install it if the current PHP version is lower than 7.2. So it should be removed before installing other packages (otherwise `composer install` fails).

> or should we file a bug report with spatie/phpunit-watch to have that project list a minimum version if one is missing?

No, it's not missing. https://github.com/spatie/phpunit-watcher/blob/master/composer.json#L19

dmsnell replied:
> Yes, composer already does that. It knows that this package is not compatible with PHP < 7.2 and refuses to install it if the current PHP version is lower than 7.2. So it should be removed before installing other packages (otherwise composer install fails).

What I meant was that maybe we can avoid hard-coding a specific dependency as we've done here. It looks like if we spit out the tree format of `composer show` we can get that information, and then process it with a small script.

```php
<?php

require 'vendor/autoload.php';

use Composer\Semver\Semver;

$PHP_VERSION = empty( getenv( 'PHP_VERSION' ) ) ? phpversion() : getenv( 'PHP_VERSION' );
$tree = json_decode( file_get_contents( 'php://stdin' ) );

function works_with_php_version( $version, $package ) {
	foreach ( $package->requires as $dep ) {
		if ( 'php' === $dep->name && ! Semver::satisfies( $version, $dep->version ) ) {
			//echo "Bailing on {$package->name} because of {$dep->version}" . PHP_EOL;
			return false;
		}

		if ( ! empty( $dep->requires ) && false === works_with_php_version( $version, $dep ) ) {
			return false;
		}
	}

	return true;
}

$unsupported = array();
foreach ( $tree->installed as $package ) {
	if ( ! works_with_php_version( $PHP_VERSION, $package ) ) {
		$unsupported[] = $package->name;
	}
}

echo implode( ',', $unsupported );
```

Invoked as `composer show --tree --format=json | php php-supported.php` this will spit out a list of packages that have dependencies which aren't supported on the running PHP version, or the version set in `PHP_VERSION` if such an ENV is set.

But it also includes another package not listed, and to the best of my ability to double-check it, it's right according to the listed minimum versions (when I set `PHP_VERSION=5.6` and run it).

```
spatie/phpunit-watcher,yoast/phpunit-polyfills
```

While maybe someone registered the wrong minimum version, or we simply haven't hit the part that breaks yet, this is why I think it'd be good at least to explore determine this through inspecting the package definitions. The list will undoubtedly change more rapidly than we come around to update it.

Edit: I should have noted that `Semver` is also already pulled in to our composer deps.

anton-vlasenko replied:
Thanks for sharing the script.
> we can avoid hard-coding a specific dependency as we've done here

Out of all installed Composer packages, this is the only package that gives trouble.
I didn't see the need to write a script for detecting such packages when I was working on the CI job, as this was the only "problematic" package.

If someone adds a composer package that is not compatible with PHP 5.6 or another Core-supported PHP version, I'd prefer that the author of that PR would be able to see the error (i.e., `composer install` should fail) instead of removing or hiding such packages with the help of a script.
Gutenberg should support the same PHP versions as Core.

If we had proper CI jobs before, `spatie/phpunit-watcher` wouldn't be merged to the repo (that's my assumption).
Removing `spatie/phpunit-watcher` before running PHPUnit tests is a workaround until that package is removed or replaced with another package that supports PHP 5.6.

`yoast/phpunit-polyfills` - I'm not sure I understand why it isn't compatible with PHP 5.6.
Please see this Packagist page:
https://packagist.org/packages/yoast/phpunit-polyfills
![Screenshot 2022-12-20 at 11 58 26](https://user-images.githubusercontent.com/43744263/208651277-e0cc5136-d06d-46d9-ae66-e2caf4af969c.png)

dmsnell replied:
> why it isn't compatible with PHP 5.6.

the dependencies of its dependencies state a minimum required version above 5.6. like I wrote above though, we may simply have not run into the places where it will break, or the package authors may have incorrectly listed their requirements.

if you run `composer show --tree` you will see where it comes in. ironically it's through `phpunit` so maybe we could safely eliminate that dependency chain as a given since we're already running.

> If someone adds a composer package that is not compatible with PHP 5.6 or another Core-supported PHP version, I'd prefer that the author of that PR would be able to see the error (i.e., composer install should fail) instead of removing or hiding such packages with the help of a script.

I totally agree! that's why I asked about removing the package. do we know if anyone would even miss the `npm run test:php:watch` scripts? I wouldn't, because I don't use that due to the way it obscures the underlying test runner and won't let me do what I want to do with `phpunit` (and it doesn't even run when I try, immediately fails with an error).

it's not the worst thing to have this hard-coded exception in the code, but I'd love to do all we can to avoid it if possible. it makes me worry that this code will end up permanent in there and we'll lose our memory of why, and then when something is broken or needing maintenance it will make that work difficult, particularly if we end up removing the referenced package.

alternatively, the worst scenario I see which seems to be quite common is that we have all this special-casing hidden deep in our layers of build scripts and then something breaks and there's no linkage between the failure and the code in the build and testing scripts that leads someone to find the problem.

anyway that's more words than it warrants, but maybe we just ask if we can get rid of `npm run test:php:watch` and not have to deal with this.

dmsnell replied:
for now, I'm not saying we can't leave in the exclusion, so don't let that block this PR. but if possible, let's rip it out, as like you said, it shouldn't have been added in the first place

if someone wants watch mode they can still run the scripts manually. they can install the package and use it.

jrfnl replied:
@dmsnell Re: the `yoast/phpunit-polyfills` package: your script must be wrong. It is 100% PHP cross-version compatible for PHP 5.4 - 8.2 and provides the PHPUnit cross-version compatibility layer, which is why it allows for a wide range of PHPUnit versions.

dmsnell replied:
@jrfnl thanks. I'm sure I'm wrong; I must be wrong. but I like being shown how and why I'm wrong, so if you think the script is wrong, well it's right there and you can point it out.

I'm clearly not an expert on this, but I am looking at the output of `composer show --tree` and for `yoast/phpunit-polyfills` it demonstrably lists dependencies for `phpunit/phpunit` which don't list 5.6 support.

as I noted above, it might make sense to special-case `phpunit` and omit checking its dependencies, _given that we are already assuming the presence of `phpunit`_, but what is wrong in the logic in the script?

I'm not just talking about the dependencies we currently have, but I'm wanting to make sure that when those dependencies change that we don't have to remember that this code exists, that it exists here, and remember to update it properly - not a good track record with relying that much on manual process.

```
yoast/phpunit-polyfills 1.0.4 Set of polyfills for changed PHPUnit functionality to allow for creating PHPUnit cross-version compatible tests
├──php >=5.4
└──phpunit/phpunit ^4.8.36 || ^5.7.21 || ^6.0 || ^7.0 || ^8.0 || ^9.0
   ├──doctrine/instantiator ^1.3.1
   │  └──php ^7.1 || ^8.0
   ├──ext-dom *
   ├──ext-json *
   ├──ext-libxml *
   ├──ext-mbstring *
   │  └──php >=7.1
   ├──ext-xml *
   ├──ext-xmlwriter *
   ├──myclabs/deep-copy ^1.10.1
   │  └──php ^7.1 || ^8.0
```

here's the truncated part of `composer show --tree` showing _some_ of the dependencies not listing PHP 5.6 support. Maybe I'm overlooking that something is optional, or that there's a way to know if these will be supported. I'm definitely relying on package maintainers to list the correct support on their packages.

jrfnl replied:
> I'm sure I'm wrong; I must be wrong. but I like being shown how and why I'm wrong, so if you think the script is wrong, well it's right there and you can point it out.

@dmsnell Sorry, I hadn't looked at the script before  - I just responded as the author of the PHPUnit Polyfills.

Looking at the script now, IMO, it is pretty clear where the error is.
1. The script will only work after a `composer install` has been run (as it needs `vendor/autoload`).
2. It will then analyse based on the dependencies which have been installed, which is **highly** dependent on which PHP version was used to run `composer install` (as there is no `composer.lock` file _exactly_ for that reason: to allow for installing the correct PHPUnit version for each PHP version).

So... say I run `composer install` on PHP 8.0, it will install PHPUnit 9.x with the associated dependencies. Neither PHPUnit 9.x, nor any of its dependencies will be compatible with PHP 5.6.

No surprise then that the Polyfills will be flagged as incompatible, but that's not a problem with the polyfills, but a problem with a logic fallacy on which the script is based....

dmsnell replied:
> Sorry, I hadn't looked at the script before - I just responded as the author of the PHPUnit Polyfills.

I definitely wish I were smart enough to not have to read the questions I respond to 🙃 

> No surprise then that the Polyfills will be flagged as incompatible, but that's not a problem with the polyfills, but a problem with a logic fallacy on which the script is based....

If I understand what you're saying though it seems to imply the script is correct, the logic in the script is correct, and I simply made the mistake of running it after installing packages with a newer version of PHP, something which wouldn't happen in the CI run.

So I think in a way by trying to show that the script is flawed you actually confirmed it's the right logic, because my goal is to determine programmatically what needs to be removed based on the running script in a way that lets us avoid depending on people manually doing all the right steps.

However…

> highly dependent on which PHP version was used to run composer install…So... say I run composer install on PHP 8.0, it will install PHPUnit 9.x with the associated dependencies. Neither PHPUnit 9.x, nor any of its dependencies will be compatible with PHP 5.6.

When I tried doing this I actually got the same result with PHP 5.6 where the `composer show --tree` output still declares `yoast/phpunit-polyfills` as unsupported, with identical output as listed above.

What I did was:
 - clone the Gutenberg repo in a `docker run php:5.6` environment
 - [install Composer through their script](https://getcomposer.org/download/)
 - run `rm composer.lock && rm -rf vendor && git reset --hard` to make sure that nothing in `vendor` was stale
 - run `./composer.phar install`
 - run `./composer.phar show --list`

```bash
~$ php --version
PHP 5.6.40 (cli) (built: Jan 23 2019 00:04:26)
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
~$ ./composer.phar --version
Composer version 2.2.18 2022-08-20 11:33:38
```

In this case though my script failed to run because of something in `symphony/polyfill` that depended on newer PHP versions. After manually removing `spatie/phpunit-watcher` it progressed further, but still failed to execute the autoloader due to an issue with unsupported `strict_types` inside the `phpunit`'s `Framework/Assets/Functions.php` module.

Maybe my version of Composer is too new? To try and see if this could be the case I tried against with Composer 1.6 from the beginning of 2018 but its output is still the same, still showing `phpunit/phpunit` pulling in dependencies that rely on PHP 7.1 and newer.

So I guess the current state is that if we want to do this programmatically the script is telling us the information from the installed packages as it should, but we need to modify it to read the `Semver` dependencies directly instead of through the autoloader, since that fails on PHP 5.6, and also ensure that we're running it on the right set of installed packages.

```php
require 'phar://composer.phar/vendor/composer/semver/src/Constraint/ConstraintInterface.php';
require 'phar://composer.phar/vendor/composer/semver/src/Constraint/MultiConstraint.php';
require 'phar://composer.phar/vendor/composer/semver/src/Constraint/Constraint.php';
require 'phar://composer.phar/vendor/composer/semver/src/Semver.php';
require 'phar://composer.phar/vendor/composer/semver/src/VersionParser.php';

use Composer\Semver\Semver;

$PHP_VERSION = empty( getenv( 'PHP_VERSION' ) ) ? phpversion() : getenv( 'PHP_VERSION' );
$tree = json_decode( file_get_contents( 'php://stdin' ) );

function works_with_php_version( $version, $package ) {
	foreach ( $package->requires as $dep ) {
		if ( 'php' === $dep->name && ! Semver::satisfies( $version, $dep->version ) ) {
			return false;
		}

		if ( ! empty( $dep->requires ) && false === works_with_php_version( $version, $dep ) ) {
			return false;
		}
	}

	return true;
}

$unsupported = array();
foreach ( $tree->installed as $package ) {
	if ( ! works_with_php_version( $PHP_VERSION, $package ) ) {
		$unsupported[] = $package->name;
	}
}

echo implode( ',', $unsupported );
```

I'd still like to better understand where I'm wrong here because I think I have tried doing what you told me I need to do but the results are the same as when I did it before.

jrfnl replied:
> If I understand what you're saying though it seems to imply the script is correct, the logic in the script is correct, and I simply made the mistake of running it after installing packages with a newer version of PHP, something which wouldn't happen in the CI run.

No, as, as you yourself just proved above.
* Your script is intended to find out which dependencies are incompatible so they can be removed **_before_** an install is run.
* Your script requires a Composer install to have been run _already_, which would fail when there are incompatible dependencies.

> Maybe my version of Composer is too new?

For an install on PHP 5.6, you can use the latest of the Composer LTS 2.2.x series.
Composer 2.3.0 and higher require PHP 7.2 (or higher).

> In this case though my script failed to run because of something in symphony/polyfill that depended on newer PHP versions. After manually removing spatie/phpunit-watcher it progressed further, but still failed to execute the autoloader due to an issue with unsupported strict_types inside the phpunit's Framework/Assets/Functions.php module.

I haven't dug in any deeper now, but I suspect this may have something to do with the default PHP version on your system not being PHP 5.6 and something somewhere along the way starting a new PHP process and using the system default PHP version instead of the PHP version you want it to use. Could that be it ?

> So I guess the current state is that if we want to do this programmatically the script is telling us the information from the installed packages as it should, but we need to modify it to read the Semver dependencies directly instead of through the autoloader, since that fails on PHP 5.6, and also ensure that we're running it on the right set of installed packages.

Possibly, and while I appreciate you are enjoying the intellectual exercise of trying to understand this. @anton-vlasenko 's two liner in the GHA script fixes this with a lot less effort and higher accuracy.

dmsnell replied:
> Your script is intended to find out which dependencies are incompatible so they can be removed before an install is run.

that's not the case, but that's how I interpreted your statement when you said "which is highly dependent on which PHP version was used to run composer install". if the running PHP version determines what packages composer will install then I took that to imply it wouldn't install packages that are listed as incompatible.

> For an install on PHP 5.6, you can use the latest of the Composer LTS 2.2.x series.

thanks. as indicated in my previous comment I did attempt this on 2.2 and 1.6 versions.

> system default PHP version instead of the PHP version you want it to use. Could that be it ?

No I don't think so, because as indicated in my previous comment, PHP 5.6 was the only version of PHP installed within the Docker container. It was both the system default and the only available version to run.

dmsnell commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,128 @@
+                  git --version
+                  svn --version
+                  composer --version
+                  locale -a
```
Is all this necessary or mostly useful while developing this branch? Same question for the Docker debug info step below.

anton-vlasenko replied:
It's not necessary, but it's worthwhile.
E.g., it helps to confirm which version of Composer and/or PHP are used when installing Composer packages and running PHPUnit.

dmsnell replied:
I'm asking because we don't currently do this for our other workflows, and it's a fair amount of noise, and some duplication (e.g. PHP and composer versions are already available through the matrix and through the PHP install step).

It's not problematic, per se, just curious why it's here as it seems like a step (to remove this debug logging) that might have been overlooked before prepping for final merge into `trunk`.

anton-vlasenko replied:
>  because we don't currently do this for our other workflows

We actually [do this](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/phpunit-tests.yml#L138).

> PHP and composer versions are already available through the matrix and through the PHP install step

Yes, they are through the matrix, but we cannot be sure that the CI jobs use these specific versions (until we log them).
Also, in the matrix, we only specify major and minor PHP versions (the same way it's implemented in Core).
The logging allows us to know the exact PHP version (including the point release) used when running particular CI jobs.

jrfnl replied:
> to know the exact PHP version (including the point release) used when running particular CI jobs.

That information is already available via the `setup-php` step.

Jonathan and me actually discussed removing a lot of this kind of debug noise from the WP Core workflows recently as the information is already available within the workflow.

anton-vlasenko replied:
> That information is already available via the setup-php step.

I've removed logging Composer and PHP versions.
cb3ee089ee66b68af28305833e5868f7d517454d
2dfb8452e972553650ac309e497cb0fee252be15

dmsnell commented on phpunit/block-supports/layout-test.php:
```diff
@@ -7,6 +7,21 @@
+	 * @var array|null
+	 */
+	private $queries;
+
```
are these (and the similar ones below) included for PHP compatibility issues?

anton-vlasenko replied:
Yes. These properties were added to ensure compatibility with PHP 8.2 and newer PHP versions.
PHP 8.2 deprecates dynamic class properties.

anton-vlasenko committed:
Rename Core to Gutenberg.

anton-vlasenko committed:
Add job's description.

anton-vlasenko committed:
Add a comment to explain why we need the /.cache folder.

anton-vlasenko committed:
Add a comment to explain why we need complicated .gitignore rules.

anton-vlasenko committed:
Fix grammar.

anton-vlasenko committed:
Revert accidental changes.

anton-vlasenko committed:
Revert accidental changes.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+name: PHP Coding Standards
+
+on:
+    push:
```
Not sure if there is a different convention in use in this repo, but the "standard" convention for Yaml files is indents of 2 spaces.

anton-vlasenko replied:
It seems that other .yaml files in this repo use four spaces indents.
E.g., https://github.com/WordPress/gutenberg/blob/trunk/.github/workflows/publish-npm-packages.yml
For this reason, I'd leave it as is.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+            # This file configures PHPCS. Changes could affect the outcome.
+            - 'phpcs.xml.dist'
+            # Changes to workflow files should always verify all workflows are successful.
+            - '.github/workflows/*.yml'
```
Please add:
```yml
  # Allow manually triggering the workflow.
  workflow_dispatch:
```

Manual triggering is not allowed by default in GHA, so needs to be explicitly permitted.

I've found this useful on numerous occasions, like when a bug was fixed in an action runner, a new Composer version was released, a new PHP patch version was released etc.

anton-vlasenko replied:
Sure. Thanks for pointing it out.
Fixed in d448a66025a438b8a318c1711159f98043053c16

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+    cancel-in-progress: true
+
+jobs:
+    # Runs PHP coding standards checks.
+    #
+    # Violations are reported inline with annotations.
+    #
+    # Performs the following steps:
+    # - Checks out the repository.
+    # - Sets up PHP.
+    # - Configures caching for PHPCS scans.
+    # - Installs Composer dependencies.
+    # - Make Composer packages available globally.
+    # - Runs PHPCS on the full codebase with warnings suppressed.
+    # - Generate a report for displaying issues as pull request annotations.
+    # - Runs PHPCS on the `tests` directory without warnings suppressed.
+    # - Generate a report for displaying `test` directory issues as pull request annotations.
+    # - Ensures version-controlled files are not modified or deleted.
```
Might be a convention in this repo, but to me this is just noise. Every step already has a name, so this is a maintenance nightmare in the making as people will forget to update this when they update the workflow steps.

anton-vlasenko replied:
This description has been copied from https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/phpunit-tests.yml#L35
Personally, I don't think it's noise, but I don't mind removing it.
Removed in 2363c402862adec91e8429a53dd18756e228b990

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+    pull_request:
+        paths:
+            # Any change to a PHP file should run checks.
+            - '**.php'
```
Is `.php` the only extension in use for PHP files ?

anton-vlasenko replied:
That's a good question.
Some PHP code could be in stored .js files
But I removed the paths section in 
https://github.com/WordPress/gutenberg/commit/1d10c2c294cad20dcea822e785bc1d88c357420c

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+            - 'release/**'
+            - 'wp/**'
+    pull_request:
+        paths:
```
I'm not sure whether limiting this by `paths` is a good idea. There is no `composer.lock` file (good, in this case), but the version constraints in the `composer.json` file are not that tightly bound, so a new release of one of the CS dependencies will now not trigger a build, while it probably should.

anton-vlasenko replied:
Fixed in 1d10c2c294cad20dcea822e785bc1d88c357420c

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+
+            - name: Run PHPCS on all Gutenberg files
+              id: phpcs-gutenberg
+              run: phpcs -n --report-full --cache=./.cache/phpcs-src.json --report-checkstyle=./.cache/phpcs-report.xml
```
```suggestion
              run: phpcs --report-full --report-checkstyle=./.cache/phpcs-report.xml
```

* `cache` directive should be moved to the ruleset.
* Why would warnings be given a free pass ?
* This command already scans the `phpunit` tests directory as well, so the separate step for scanning the `phpunit` tests directory is redundant.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,101 @@
+              if: ${{ always() && steps.phpcs-gutenberg.outcome == 'failure' }}
+              run: cs2pr ./.cache/phpcs-report.xml
+
+            - name: Check test suite files for warnings
+              id: phpcs-tests
+              run: phpcs phpunit --report-full --cache=./.cache/phpcs-tests.json --report-checkstyle=./.cache/phpcs-tests-report.xml
+
+            - name: Show test suite scan results in PR
+              if: ${{ always() && steps.phpcs-tests.outcome == 'failure' }}
+              run: cs2pr ./.cache/phpcs-tests-report.xml
```
Should be removed as per the above, unless there is a good reason to allow warnings in `src`, but not in `tests`.

If that's the case, the first CS run needs to be updated to exclude the `phpunit` directory as that's now being scanned twice.

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,144 @@
+        branches:
+            - trunk
+            - 'release/**'
+            - 'wp/**'
```
Same remark about adding the `workflow_dispatch`

anton-vlasenko replied:
Fixed in https://github.com/WordPress/gutenberg/commit/d448a66025a438b8a318c1711159f98043053c16

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,144 @@
+    cancel-in-progress: true
+
+jobs:
+    # Runs the PHPUnit tests for Gutenberg.
+    #
+    # Performs the following steps:
+    # - Sets environment variables.
+    # - Checks out the repository.
+    # - Sets up Node.js.
+    # - Sets up PHP.
+    # - Installs Composer dependencies.
+    # - Installs npm dependencies
+    # - Logs general debug information about the runner.
+    # - Logs Docker debug information (about the Docker installation within the runner).
+    # - Starts the wp-env Docker containers.
+    # - Logs the running Docker containers.
+    # - Logs debug information about what's installed within the wp-env Docker containers.
+    # - Install WordPress and Gutenberg within the Docker container.
+    # - Run the PHPUnit tests.
```
Same remark about not duplicating information already available.

anton-vlasenko replied:
Removed in https://github.com/WordPress/gutenberg/commit/2363c402862adec91e8429a53dd18756e228b990

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,144 @@
+    # - Run the PHPUnit tests.
+    test-php:
+        name: ${{ matrix.php }}${{ matrix.multisite && ' multisite' || '' }} on ${{ matrix.os }}
+        runs-on: ${{ matrix.os }}
```
Why have the `os` as a matrix value and in the job name if only one OS is being used ?

anton-vlasenko replied:
This code has been copied from the corresponding Core CI job.
I agree. It makes no sense to have it as a matrix value.
Fixed in 8210cb28e67787e8172ea686dc2ef1825eb5e1ae

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,144 @@
+        timeout-minutes: 20
+        if: ${{ github.repository == 'WordPress/gutenberg' || github.event_name == 'pull_request' }}
+        strategy:
+            fail-fast: false
```
Is there a reason to not fail fast ? Removing this saves a lot of resources.

anton-vlasenko replied:
My reasoning for setting it to `false` is the following:
It's convenient to see all CI jobs on which tests are failing.
Often unit tests fail on some PHP versions while passing on others.
If the `fail-fast` option is true, GitHub will show only the first failed job.

jrfnl replied:
I totally get the convenience argument.
Having said that, in my experience, more often than not, when there is a tests failure, it fails on all PHP versions.
Of course, there are times when it is PHP version specific and in those cases, the runtime of the test runs will end up deciding when the build fails and what shows, which is not always helpful.

I'm not sure what type of GH account the WP organisation has, but I find that in most cases these (counter-)arguments apply:
* If it's a free account: be kind to GH, which is shouldering the costs, by only using the resources we need.
* If it's a paid account: be kind to accounting as every build counts towards the minutes billed.
* In both cases: be kind to the environment and don't use more server resources than are strictly needed.

Let's hear from some more people/opinions ?

anton-vlasenko replied:
I've found an interesting [Trac](https://core.trac.wordpress.org/ticket/52612) ticket.
It explains why the `fail-fast` option was disabled in Core.
I thought Gutenberg should profess the same principles.

> If it's a free account: be kind to GH, which is shouldering the costs, by only using the resources we need.

As far as I understand this repo is far from hitting the usage limits, as each job is allowed [to run for 6 hours](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits).

> Let's hear from some more people/opinions ?

Yes, it would be interesting to hear more opinions on this matter.
Please feel free to tag anyone who could share an opinion on this matter.
Thanks!
Speaking of me, I still see value in having it disabled because of the reasons above.

anton-vlasenko replied:
I've found an interesting [Trac](https://core.trac.wordpress.org/ticket/52612) ticket.
It explains why the `fail-fast` option was disabled in Core.
I thought Gutenberg should profess the same principles.

> Let's hear from some more people/opinions ?

Yes, we can do it.
But if it takes long to get someone else to review this, I'd prefer to enable fast failing now. It can be disabled later if it causes issues (as it did in Core).

anton-vlasenko replied:
I've enabled fast-fail as there were no more comments.
9452534594cc5e60b9b35b83e197fdd6a251c317

ockham replied:
A bit late to this, but I agree with fast-failing for the reasons that y'all mentioned, especially:

> more often than not, when there is a tests failure, it fails on all PHP versions

...and the resources that fast failing allows us to save.

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,144 @@
+        strategy:
+            fail-fast: false
+            matrix:
+                php:
+                    [
+                        '5.6',
+                        '7.0',
+                        '7.1',
+                        '7.2',
+                        '7.3',
+                        '7.4',
+                        '8.0',
+                        '8.1',
+                        '8.2',
+                    ]
```
Either specify it like so:
```yaml
php: ['5.6', '7.0', '7.1', '7.2', '7.3', '7.4', '8.0', '8.1', '8.2']
```
or like so:
```yaml
php:
  - '5.6'
  - '7.0'
  - '7.1'
  - '7.2'
  - '7.3'
  - '7.4'
  - '8.0'
  - '8.1'
  - '8.2'  
```

This is Yaml, not PHP.

dmsnell replied:
I agree with @jrfnl on this, but…

> This is Yaml, not PHP.

I don't see the reason to add snark in this comment. The code as written parses as expected and is valid YAML. We don't have to claim that @anton-vlasenko was unaware of which language is being written or put anyone down when their style is different than your style or even when we find stylistic differences in a project and a code submission.

jrfnl replied:
Sorry if you interpreted my comment as snarky, definitely wasn't meant as such. It was just meant to convey that I understand where the inclination to write it like that comes from, nothing more.

dmsnell replied:
> I understand where the inclination to write it like that comes from

I'm not the Gutenberg maintainer, so this isn't coming from any authority, in the way that you have expressed multiple times having authority yourself, but what I share here is meant to be a reminder of a way that help avoid these kinds of misunderstandings: we believe we understand the intentions and inclinations behind someone else's behaviors more than we really do understand, and leaning on that assumption is the source of many needless problems.

when we start by assuming we understand someone else's motivations it just about guarantees friction and trouble. we're not giving those people a chance to share their motivations and we're often starting by assuming they are wrong (which also means we're not giving ourselves an opportunity to learn from something that seems wrong at our quick glance).

it's fine to notice a style issue like this and share [as you did in the next comment](https://github.com/WordPress/gutenberg/pull/46510#discussion_r1053847254) but it's a big leap to start assuming we know what's inside someone else's head and then to assert it publicly, and then tell them that they are wrong because of the way we know they were thinking.

I've far too often had to deal with big problems that arise because of a "fix" that was made due to something looking out of place but being there for a reason, and the fixer comes in and changes it because they assume it needs to be removed, assume it was a mistake the author made. obviously we come across oddities and quirks all the time, and in most cases those are good points to revise code that was incidentally awkward, but I believe we can build stronger relationships if we start by asking _why_ something is different than we expect rather than telling someone "I know why you did this and it's wrong," especially if those declarations do end up being wrong and the cause for real failures that do bring a lot of frustration.

you bring a lot of good value to a review like this; I want to acknowledge that and appreciate your time and attention there. just want to ask at the same time if we can bring positive assumptions to the discussion, like assuming that people are smarter than we give them credit for, or have reasons behind the choices they make.

jrfnl replied:
Please take your own advise.

jrfnl replied:
And just to be crystal clear what I mean: you seem to be making all sorts of assumptions about me and are interpreting my comments with that in your mind and looking for meaning behind the words, where there is none. My comments are plain and should be taken at face-value. Nothing more. I can't help it if someone reads something into them what isn't there.

anton-vlasenko replied:
I don't see @jrfnl 's comment as snarky either. 
I sincerely appreciate your expertise and your code review, @jrfnl.
Thank you!

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,144 @@
+
+            - name: Running single site unit tests
+              run: npm run test:unit:php
+              if: ${{ ! matrix.multisite }}
```
Nitpick: in most repos, the convention is to have the `if:` above the `run:`

anton-vlasenko replied:
Fixed in 9df15b6cb010f0c071ae3da8856dec1d03e519df

jrfnl commented on packages/env/lib/build-docker-compose-config.js:
```diff
@@ -183,7 +183,7 @@ module.exports = function buildDockerComposeConfig( config ) {
 		phpunitTag = '7' + phpunitPhpVersion;
 	} else if ( [ '7.2', '7.3', '7.4' ].indexOf( testsPhpVersion ) >= 0 ) {
 		phpunitTag = '8' + phpunitPhpVersion;
-	} else if ( testsPhpVersion === '8.0' ) {
+	} else if ( [ '8.0', '8.1', '8.2' ].indexOf( testsPhpVersion ) >= 0 ) {
 		phpunitTag = '9' + phpunitPhpVersion;
```
This condition block seems wrong/out of date.
PHP 7.2 needs PHPUnit 8.x.
PHP 7.3 and higher should use PHPUnit 9.x

Also see: https://phpunit.de/supported-versions.html

anton-vlasenko replied:
Yes, I saw that page.
But I was primarily focused on making wp-env work with PHP 8.2.
This slipped under the radar.
Thanks for pointing out!
Fixed in d7195bf64a5730fa4a5069cc02232e18edf6f713 and 26687328057b3c6052e65d7d1415ce2c20a9cde3

jrfnl replied:
🙌🏻

jrfnl reviewed (changes requested):
Thanks for the ping @anton-vlasenko and for setting this up! Great step forward.

I've left a number of comments inline. Take from it what you will.

Other than that, I truly don't understand why NPM and Docker are needed. Seems like a lot of overkill and a roundabout way to do things, which can just be done using a PHP based workflow. Then again, I'm not intimately familiar with GB, so there may be good reasons for it, in which case, it might be a good idea to document those.

anton-vlasenko committed:
Allow manually triggering the workflows.
Props to @jrfnl for suggesting it.

anton-vlasenko committed:
Remove description.

anton-vlasenko committed:
Let's not limit the job by path.

anton-vlasenko committed:
PHP 7.3 and 7.4 should use PHPUnit 9.
Props to @jrfnl for finding this bug.

anton-vlasenko committed:
1. Fix yaml synthax.
2. Don't put the OS version in the matrix.

anton-vlasenko committed:
In most repos, the convention is to have the if: above the run:.
 Props to @jrfnl.

anton-vlasenko committed:
Don't log the PHP version; this information is available in the "Set up PHP" step.

anton-vlasenko committed:
Don't log the Composer version; this information is available in the "Set up PHP" step.

anton-vlasenko committed:
1. Don't scan the unit tests twice.
2. Move the cache option to the phpcs.xml.dist

anton-vlasenko committed:
Fix the cs2pr report location.

anton-vlasenko committed:
Update the .gitignore rule description.

anton-vlasenko commented:
Thanks for the review, @jrfnl.
I appreciate the improvements you proposed (especially those related to wrong PHPUnit versions and the `phpcs` cache).

> Other than that, I truly don't understand why NPM and Docker are needed. Seems like a lot of overkill and a roundabout way to do things, which can just be done using a PHP based workflow. Then again, I'm not intimately familiar with GB, so there may be good reasons for it, in which case, it might be a good idea to document those.

I see your point.

My reasoning for using Docker and `wp-env` is the following:
1. `wp-env` is simple to install and requires almost no configuration.
2. Gutenberg community maintains the `wp-env` package, so the CI jobs will require less maintenance when the new versions of Gutenberg/WordPress get released.
3. Running Gutenberg unit tests natively requires more hassle than running WordPress unit tests, as Gutenberg needs WordPress to be installed and configured. `wp-env` significantly simplifies this process.
4. Docker runs much better on Linux compared to macOS and Windows. The overhead is very little, and the speed is native-like.
5. Should the community need to use some other WordPress plugins or themes when running the unit tests, `wp-env` can quickly achieve that.

I have nothing against running the unit tests native. I just see more benefits in enabling testing on all PHP versions sooner than later to prevent merging incompatible/sub-optimal PHP code to the repo.

jrfnl commented:
> I have nothing against running the unit tests native. I just see more benefits in enabling testing on all PHP versions sooner than later to prevent merging incompatible/sub-optimal PHP code to the repo.

I fully agree with getting this merged sooner rather than later. This can always be iterated on at a future point in time. 👍🏻 

Re: Docker - I have the same objections to WP Core running the tests via Docker.
1. The Docker images are set up to be "ideal" environments, which is great for development, but not for testing.
2. Ideal environments are not representative of the real world environments WP runs on.
3. We need a much higher variation of test environments:
    - Testing with various PHP extensions disabled (like Sodium, Imagick, GD, but there are definitely more to look at).
    - Testing with different versions of PHP extensions, especially the PECL ones, like Imagick, which may not get updated alongside PHP.
    - Testing with a better range of PHP/MySQL combinations.
    - Testing with specific `ini` directives set to non-standard values (for those `ini` directives which are not runtime changable from within the tests).
    - And yes, we should also be testing against Windows server environments.
    - etc

There are currently tests in the WP Core test suite which are **_never_** run in CI due to the "ideal" Docker environment setup. This is not good.

Let's be realistic here: we are not building software which will only run in an environment over which we have full control, like in-company software.
We are building software which powers 40% of the internet and is run on an extremely wide range of system configurations, so our test setup should reflect that.

Note: For Core this is an ongoing discussion and something I would very much like to change (definitely outside the scope of this PR though).

For GB, as this is a new test workflow, you have the opportunity to do _better_ than Core, but I fully respect it if you choose to start with this and iterate on it once this is changed in WP Core.

anton-vlasenko committed:
Enable fast failing to save Github resources as proposed here: https://github.com/WordPress/gutenberg/pull/46510/files#r1055900997

anton-vlasenko commented:
Thanks for listing all the issues with the current testing environment.
I wasn't aware of that.
I agree that the more testing environments there are, the better.

> For GB, as this is a new test workflow, you have the opportunity to do _better_ than Core, but I fully respect it if you choose to start with this and iterate on it once this is 
changed in WP Core.

I cannot tell for sure that I will backport changes from Core to Gutenberg once the Core testing environment gets changed.
But I'm all for backporting the updated Core testing workflow to Gutenberg once it's merged.


I'd appreciate it if you could approve this PR, @jrfnl.
I've [enabled](https://github.com/WordPress/gutenberg/commit/9452534594cc5e60b9b35b83e197fdd6a251c317) the fail-fast option in the matrix job.

jrfnl commented on .github/workflows/coding-standards.yml:
```diff
@@ -0,0 +1,74 @@
+              if: ${{ always() && steps.phpcs-gutenberg.outcome == 'failure' }}
+              run: cs2pr ./.cache/phpcs-report.xml
+
+            - name: Show test suite scan results in PR
+              if: ${{ always() && steps.phpcs-tests.outcome == 'failure' }}
+              run: cs2pr ./.cache/phpcs-report.xml
```
As the separate PHPCS run for the tests has been removed, this step should be removed as well.

anton-vlasenko replied:
Great catch.
Fixed in 48da919c8f2b0ad57c817a72998f45e2c56cf7ba

jrfnl commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -0,0 +1,125 @@
+            ##
+            - name: Set up PHP
+              uses: shivammathur/setup-php@1a18b2267f80291a81ca1d33e7c851fe09e7dfc4 # v2.22.0
+              with:
+                  php-version: '${{ matrix.php }}'
+                  coverage: none
```
Not sure how relevant this is, what with the tests being run via Docker, but Setup PHP defaults to using the `production` php.ini file, which hides deprecation notices and such (while we definitely do want to see those for the test runs).

To change this, add either of the below to the `with:` key:
```
ini-values: zend.assertions=1, error_reporting=-1, display_errors=On, log_errors_max_len=0
# or
ini-file: development
```

anton-vlasenko replied:
That's good to know.
Fixed in eaef3343f85faa8f50f7465eb70105cac961d0ca

jrfnl replied:
Looks like you "fixed" this in the coding standards repo (where it isn't needed) and not in the test workflow, where it **is** needed... ?

jrfnl reviewed (changes requested):
Two more small remarks, but other than that, all looks good now.

anton-vlasenko committed:
As the separate PHPCS run for the tests has been removed, this step should be removed as well.
Props to @jrfnl.

anton-vlasenko committed:
Use the development php.ini file (to make any errors more obvious).
Props to @jrfnl.

anton-vlasenko committed:
Use the development php.ini file (to make any errors more obvious).
Props to @jrfnl.

anton-vlasenko committed:
Revert "Use the development php.ini file (to make any errors more obvious)."

This reverts commit eaef3343f85faa8f50f7465eb70105cac961d0ca.

jrfnl commented:
@anton-vlasenko Quick question: is there a Dependabot config in the repo to automatically update the predefined action runners ? If not, I'd strongly advise to double-check the used versions as quite a few look out of date to me.

* `actions/checkout` uses version `3.1.0`, while IIRC the latest release is `3.3.0`.
* `shivammathur/setup-php` uses version `2.22.0`, while IIRC the latest release is `2.23.0`.
* `actions/cache` uses version `3.0.11`, while IIRC the latest release is `3.2.3`.
...etc...

jrfnl reviewed (approved):
Thanks for making those fixes @anton-vlasenko !

I've confirmed my concerns where addressed.
I have left [one more question](https://github.com/WordPress/gutenberg/pull/46510#issuecomment-1380786866), which you may still want to address, but other than that, all looks good to me.

anton-vlasenko committed:
Update the changelog.
Props to @jrfnl for identifying an issue with PHP 7.4 and 7.4.

# Conflicts:
#	packages/env/CHANGELOG.md

anton-vlasenko committed:
Move the unit tests to unit-test.yml.
This is needed because:
1. unit-php job has to be preserved as it's set as required in the GH settings.
2. unit-php job now depends on the test-php job.

anton-vlasenko commented:
> @anton-vlasenko Quick question: is there a Dependabot config in the repo to automatically update the predefined action runners ? If not, I'd strongly advise to double-check the used versions as quite a few look out of date to me.
> 
> * `actions/checkout` uses version `3.1.0`, while IIRC the latest release is `3.3.0`.
> * `shivammathur/setup-php` uses version `2.22.0`, while IIRC the latest release is `2.23.0`.
> * `actions/cache` uses version `3.0.11`, while IIRC the latest release is `3.2.3`.
>   ...etc...

@jrfnl Yes, there is a Dependabot config in the repo.
Hopefully it will take care of updating the versions.
https://github.com/WordPress/gutenberg/blob/trunk/.github/dependabot.yml

anton-vlasenko committed:
Change the description.

anton-vlasenko committed:
Rename the job.
